packages/devkit/src/utils/mapped-graph.ts

```diff
@@ -0,0 +1,40 @@
+import {
+  ProjectGraph,
+  ProjectGraphProjectNode,
+} from 'nx/src/config/project-graph';
+
+export type MappedProjectGraph<T = any> = ProjectGraph<T> & {
+  allFiles: Record<string, string>;
+};
+
+export function removeExt(file: string): string {
+  return file.replace(/(?<!(^|\/))\.[^/.]+$/, '');
+}
+
+/**
+ * Maps the project graph to a format that makes it easier to find the project
+ * based on the file path.
+ * @param projectGraph
+ * @returns
+ */
+export function mapProjectGraphFiles<T>(
+  projectGraph: ProjectGraph<T>
+): MappedProjectGraph | null {
+  if (!projectGraph) {
+    return null;
+  }
+  const allFiles: Record<string, string> = {};
+  Object.entries(
+    projectGraph.nodes as Record<string, ProjectGraphProjectNode>
+  ).forEach(([name, node]) => {
+    node.data.files.forEach(({ file }) => {
+      const fileName = removeExt(file);
```

Comment:
This is a weird way to make a key. Why is it not the `dirname`?

Reply:
Because our imports get resolved to `/path/to/folder/file` and lookup gets invoked 1000s of times during the lint run.
We need a lookup that's as fast as possible.